### PR TITLE
Courses

### PR DESCRIPTION
**Pull request by weilingooi**
Courses

### CODE REVIEW

**Comment by weilingooi** [comment]
ready

File: app/assets/stylesheets/courses.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Courses controller here.
+// They will automatically be included in application.css.
```
Comment (ngocvt-0484) [inline app/assets/stylesheets/courses.scss]: delete empty file

File: app/controllers/courses_controller.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
+class CoursesController < ApplicationController
+  before_action :set_course, only: [:show, :edit, :update, :destroy]
+  before_action :authenticate_user! , only: [:edit, :update , :destroy]
```
Comment (ngocvt-0484) [inline app/controllers/courses_controller.rb]:
```suggestion
  before_action :authenticate_user! 
  before_action :set_course, only: [:show, :edit, :update, :destroy]
```

File: app/controllers/courses_controller.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  before_action :check_permitted_or_not , only: [:edit , :destroy]
+  before_action :authenticate_user! , only: [:new , :create , :edit]
```
Comment (ngocvt-0484) [inline app/controllers/courses_controller.rb]: delete this line

File: app/controllers/courses_controller.rb [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+  def index
+    @courses = Course.all
+  end
+
+  def show; end
+
+  def check_permitted_manage_courses_or_not
+    if current_user && current_user.usertype == 1
```
Comment (ngocvt-0484) [inline app/controllers/courses_controller.rb]:
```suggestion
    redirect_to courses_path if current_user.usertype == 1
```

File: app/controllers/courses_controller.rb [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+  def check_permitted_or_not
+    if current_user
+      if @course.try(:user).try(:id) != current_user.id
+        redirect_to courses_path()
+      end
+    end
+  end
+
+  def authenticate_user!
```
Comment (ngocvt-0484) [inline app/controllers/courses_controller.rb:78]: Why don't you login and then go to this page

remove this function plz

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -25,8 +33,52 @@ def create
...
+  end
+
+
   private
   def user_params
     params.require(:user).permit :name, :email, :password, :password_confirmation
   end
+
+  def set_user
+    @user = User.find(params[id])
```
Comment (ngocvt-0484) [inline app/controllers/users_controller.rb]:
```suggestion
    @user = User.find_by(id: params[:id])
```

File: app/helpers/courses_helper.rb [hunk 7]
```diff
@@ -0,0 +1,2 @@
+module CoursesHelper
```
Comment (ngocvt-0484) [inline app/helpers/courses_helper.rb]: delete it

File: app/controllers/courses_controller.rb (hunk 5 above)
Comment (weilingooi) [reply app/controllers/courses_controller.rb:67]: I get error for it

Comment (weilingooi) [reply app/controllers/courses_controller.rb:78]: > Why don't you login and then go to this page
> 
> remove this function plz



I got it error

**Comment by weilingooi** [comment]
ready

File: app/controllers/courses_controller.rb [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  before_action :authenticate_user!
+  before_action :check_permitted_or_not , only: [:edit , :destroy]
+
+  def index
+    @courses = Course.all
+  end
+
+  def show; end
+
+  def check_permitted_manage_courses_or_not
```
Comment (vanvtt-0952) [inline app/controllers/courses_controller.rb]: Scope private for this method

File: app/controllers/courses_controller.rb [hunk 9]
```diff
@@ -0,0 +1,89 @@
+class CoursesController < ApplicationController
+  before_action :set_course, only: [:show, :edit, :update, :destroy]
+  before_action :authenticate_user!
+  before_action :check_permitted_or_not , only: [:edit , :destroy]
+
+  def index
+    @courses = Course.all
```
Comment (vanvtt-0952) [inline app/controllers/courses_controller.rb]: you should define `sort for name of course` and `paginate` for courses
example:
- scope for name
in  Course model
`scope :sort_by_name, , ->{order name: :desc}`
- Call scope and paginate
` @courses = Course.all` change to ` @courses = Course.sort_by_name..paginate page: 10`

File: app/controllers/courses_controller.rb [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+        format.json { render :show, status: :ok, location: @course }
+      else
+        format.html { render :edit }
+        format.json { render json: @course.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  def destroy
+    @course.destroy
```
Comment (vanvtt-0952) [inline app/controllers/courses_controller.rb]: handling when destroy fail

**Comment by weilingooi** [comment]
ready

File: app/controllers/courses_controller.rb [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
```
Comment (vanvtt-0952) [inline app/controllers/courses_controller.rb]: remove this line

File: app/controllers/courses_controller.rb [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+  end
+
+  private
+
+  def course_params
+    params.require(:course).permit(:title,:user_id)
+  end
+
+  def set_course
+    @course = Course.find_by(params[:user_id])
```
Comment (vanvtt-0952) [inline app/controllers/courses_controller.rb:63]: plz, handling when course is nil
add code 
`return if @course`
`flash[:warning] = "Course not found"`
`redirect_to root_path`

File: app/controllers/users_controller.rb [hunk 13]
```diff
@@ -25,8 +33,51 @@ def create
...
+        format.json { render :show, status: :ok, location: @user }
+      else
+        format.html { render :edit }
+        format.json { render json: @user.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  def destroy
+    @user.destroy
```
Comment (vanvtt-0952) [inline app/controllers/users_controller.rb]: handle when destroy fail

File: app/controllers/users_controller.rb [hunk 14]
```diff
@@ -25,8 +33,51 @@ def create
...
+    respond_to do |format|
+      format.html { redirect_to users_url, notice: t("controller.users.delete") }
+      format.json { head :no_content }
+    end
+  end
+
+  def check_permitted_or_not
```
Comment (vanvtt-0952) [inline app/controllers/users_controller.rb:69]: scope private for this method

File: app/controllers/users_controller.rb [hunk 15]
```diff
@@ -25,8 +33,51 @@ def create
...
+    if @user.id != current_user.id
+      redirect_to users_path
+    end
+  end
+
+  def authenticate_user!
```
Comment (vanvtt-0952) [inline app/controllers/users_controller.rb:75]: scope private for this method

File: app/controllers/users_controller.rb [hunk 16]
```diff
@@ -25,8 +33,51 @@ def create
...
+  end
+
   private
   def user_params
     params.require(:user).permit :name, :email, :password, :password_confirmation
   end
-end
+
+  def set_user
+    @user = User.find_by(id: params[:id])
```
Comment (vanvtt-0952) [inline app/controllers/users_controller.rb]: handle user is nil IN HERE

**Comment by weilingooi** [comment]
ready

File: app/controllers/courses_controller.rb [hunk 17]
```diff
@@ -0,0 +1,95 @@
+class CoursesController < ApplicationController
+  before_action :set_course, only: [:show, :edit, :update, :destroy]
```
Comment (saobv-2223) [inline app/controllers/courses_controller.rb]:
```suggestion
  before_action :set_course, except: %i(new create show)
```

File: app/controllers/courses_controller.rb [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+      else
+        format.html {redirect_to courses_path, notice: t("controller.course.not_destroy")}
+        format.json { head :no_content }
+    end
+  end
+
+  private
+
+  def course_params
+    params.require(:course).permit(:title,:user_id)
```
Comment (saobv-2223) [inline app/controllers/courses_controller.rb]: remove ()

File: app/controllers/courses_controller.rb [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+    else
+      render json: { status: 401 }
+    end
+  end
+  
+  def check_permitted_manage_courses_or_not
+    redirect_to courses_path if current_user.usertype == 1
+      redirect_to courses_path
+    end
+  end
```
Comment (saobv-2223) [inline app/controllers/courses_controller.rb]:
```suggestion
end

```

File: app/controllers/users_controller.rb [hunk 20]
```diff
@@ -1,4 +1,10 @@
 class UsersController < ApplicationController
+  before_action :set_user, only: [:show, :edit, :update, :destroy]
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: image

File: app/controllers/users_controller.rb [hunk 21]
```diff
@@ -25,8 +33,54 @@ def create
...
+
   private
   def user_params
     params.require(:user).permit :name, :email, :password, :password_confirmation
   end
-end
+
+  def set_user
+    @user = User.find_by(id: params[:id])
+    if return @user
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]:
```suggestion
    @user = User.find_by id: params[:id]
    return if @user
```

File: app/controllers/users_controller.rb [hunk 22]
```diff
@@ -25,8 +33,54 @@ def create
...
+      session[:user_id] = user.id
+      render json: {
+        status: :created,
+        logged_in: true,
+        user: user
+      }
+    else
+      render json: { status: 401 }
+    end
+  end
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]:
```suggestion
  end
  
```

File: app/views/courses/_course.html.erb [hunk 23]
```diff
@@ -0,0 +1,15 @@
...
+    <%= course.status %>
+  </p>
+
+  <%= link_to t("courses.course.view_detail"), course %> 
+
+  <% if current_user.admin? %>
+    <%= link_to t("courses.course.update_course"), edit_course_path(course) %> |
+    <%= link_to t("courses.course.delete_course"), course, method: :delete,data: { confirm: "You sure?" } %>
+  <% end %>
+</li>
```
Comment (saobv-2223) [inline app/views/courses/_course.html.erb]:
```suggestion
</li>

```

File: app/views/courses/_form.html.erb [hunk 24]
```diff
@@ -0,0 +1,24 @@
+<%= form_for(@course) do |f| %>
```
Comment (saobv-2223) [inline app/views/courses/_form.html.erb]: remove ()

**Comment by weilingooi** [comment]
ready

**Pull request merged**